### PR TITLE
Map `snabb-lwaftr' argv[0] to programs.lwaftr

### PR DESCRIPTION
**Pull request by wingo**
Rework the parsing of the Snabb command line to allow us to ship
a nicely named snabb-lwaftr binary, but have it map internally to the
also nicely-named 'programs.lwaftr' module tree.  Gets rid of the
nasty "snabb" name duplication in the module tree.

### CODE REVIEW

**Comment by lukego** [comment]
Interesting! I thought we had already adopted this behavior but looking at the diff it doesn't seem so :).

File: src/core/main.lua [hunk 1]
```diff
@@ -33,39 +33,34 @@ function main ()
...
-function select_program (program, args)
-   if program:match("^snabb_") then
-      return program:gsub("^snabb_", "")
+-- Take the program name from the first argument, unless the first
+-- argument is "snabb", in which case pop it off, handle any options
+-- passed to snabb itself, and use the next argument.
+function select_program (args)
+   local program = programname(table.remove(args, 1))
+   if program == 'snabb' then
+      while #args > 0 and args[1]:match('^-') do
```
Comment (eugeneia) [inline src/core/main.lua]: Why is there a while loop here? I see three cases and no loop:

* `-h`/`--help`—print usage
* `-x`/`--foo`—error
* `foo`—program

**Comment by eugeneia** [comment]
@lukego I think we are thinking of #640 .

File: src/core/main.lua (hunk 1 above)
Comment (wingo) [inline src/core/main.lua]: It's for if Snabb gets any more options.  LMK if you want s/while/if/ and s/do/then/; it will save a character ;)  By only checking for --help / -h the existing code won't detect invalid options, instead treating them as program names; for that reason I thought being more general would be more reliable.

Comment (eugeneia) [inline src/core/main.lua]: Its a confusing piece of code that I had to read a couple times to know what its doing. I expected a case where it actually loops but since `usage` always exits it never does.

> It's for if Snabb gets any more options.

That explains the motivation behind this and I get it. I recommend avoiding coding for the future, as its confusing for readers.

Comment (wingo) [inline src/core/main.lua]: Excuse my denseness, I have trouble knowing what your comments mean for me.  I prefer to leave it in, you prefer to leave it out, however what I don't know is whether you as reviewer require me to change the word "while" to "if" and "do" to "then".  If so please let me know.  Tx :)

Comment (eugeneia) [inline src/core/main.lua]: Yeah, the two of us tend to talk past each other. FIXME ;-)

I generally do not set the requirements for accepting a PR, so I will usually not tell you or anybody else to “do X”. Instead, the project is developed in consensus of the contributors, and I can not truthfully represent that consensus on my own. So what I do is point out issues I see and give arguments as to why I see them as issues. E.g. I supply a single “vote” to the consensus, which is not itself a veto. Deciding what to do with my view is up to you. If you are convinced by my argument you might change the code or try to avoid similar issues in the future. If you are not convinced I would ideally get back an argument that explains why. Unless anybody else takes issue with the code I will merge it, and I will usually wait for a while to give others the chance to look at the PR, discuss it and maybe agree on changes.

So in this particular case I am simply making you aware of the fact, that as a reader of the code I found a particular part of it confusing (its just feedback). I couldn't tell from your reply if you understood my confusion, so I explained it in depth in the hope that you become aware of the general issue. (I also pointed out which part of your comment I found helpful, to give you a feel of what I would like to hear.) My goal (and I think this benefits everybody) is to reduce workload by increasing empathy with the consumers of code: when code is harder that necessary to read for reviewers, it will be more work for all. So in order to reduce that work I am letting you know how it would be easier for me (the reviewer in this case). Same goes for comments I guess, we need to learn what we actually want to hear from each other.

So to sum it up: You don't need to ask me what to do because I won't tell you what to do. It's your call, not mine. What I will tell you is my perspective on your code/comments/... and maybe you can use this second pair of eyes  to make it better.

Comment (lukego) [inline src/core/main.lua]: @eugeneia I think there are two different hats that one can wear when reviewing code:

Community member: just giving feedback, nothing is binding, no responses are required.

Upstream maintainer: sole person responsible for accepting/merging the branch or spelling out explicitly what needs to be done before that can happen.

If you are the person who is resposible for upstreaming the PR and you don't press "Merge" then this is implicitly strong feedback "this code is not currently acceptable to me and so I have declined to merge it." In that case the contributor needs to clearly understand what are the deal-breakers so that they can fix them and the branch can be merged. The maintainer might say "You have to change A to B and remove C" or "we need to get an ack from @somebody or @someonelse" and then once that is done the contributor will expect the branch to merge.

This is why I have been stressing the importance of unambiguously dispatching incoming PRs between upstream maintainers. Then everybody knows who is wearing which hat and what is the path forward.

Overall there will be agreement between O(log(numberofmaintainers)) people who each merge the change onto their upstream branch as it makes its way to master.

**Comment by someonelse** [comment]
My user name is @someonelse, I don't think I was meant to be included in this coding haha.

**Comment by lukego** [comment]
@someonelse thank you for letting me borrow your name for the purpose of that example :-)

**Comment by eugeneia** [comment]
Merged into `max-next`.

**Comment by wingo** [comment]
Thanks!
